### PR TITLE
Jwong/switch to array and polymer notify

### PR DESCRIPTION
**Pull request by jwong-d2l**

### CODE REVIEW

File: demo/index.html [hunk 1]
```diff
@@ -20,7 +20,7 @@
       <h3>Basic activity-exemptions demo</h3>
       <demo-snippet>
         <template>
-          <activity-exemptions test-data host-url="" org-unit-id="" activity-id=""></activity-exemptions>
+          <activity-exemptions host-url="qa10768377g.bspc.com" org-unit-id="129502" activity-id="https://ids.brightspace.com/activities/dropbox/6E09A215-8089-48F3-B66B-32802692AF9B-3481"></activity-exemptions>
```
Comment (scasagrande-d2l) [inline demo/index.html]: Should we be commiting this?

File: test.html [hunk 2]
```diff
@@ -0,0 +1,29 @@
+<!doctype html>
```
Comment (scasagrande-d2l) [inline test.html]: What is this file for?

File: activity-exemptions.html [hunk 3]
```diff
@@ -35,7 +35,7 @@
 		<iron-ajax
 		id="classlistRequest"
 		url="[[classlistUrl]]"
-		headers='{ "Authorization": "Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiIsImtpZCI6IjNhOTJkMjRjLWU2ZWEtNGYzYy1iNGFhLTAzYzViZTAzN2Y1NCJ9.eyJzdWIiOiIxNjkiLCJ0ZW5hbnRpZCI6IjZiZDFjZDc3LWJmOTMtNGEzZi05Mzg5LTcxMzFmY2YyOWE2NiIsInNjb3BlIjoiKjoqOioiLCJqdGkiOiIxYzExYjhjMi0yMGZiLTRmMTgtYjBjOS0yZmY5NDY3YzRhNWYiLCJpc3MiOiJodHRwczovL2FwaS5icmlnaHRzcGFjZS5jb20vYXV0aCIsImF1ZCI6Imh0dHBzOi8vYXBpLmJyaWdodHNwYWNlLmNvbS9hdXRoL3Rva2VuIiwiZXhwIjoxNTA0NzM1NTU5LCJuYmYiOjE1MDQ3MzE5NTl9.Ct6Oeji3llJiB3oKUcZ0tQRjlMIahWRmfRn4UQBayo8wtUcOPtj8dwMsvsowH9BbQT2s-u3Md5wDTCDRu18Dzn3RtCR1zm6YSOtFMzuDl-T-BewaveUT15-8TBJPG9hVvKLSPaPDbJQ67R_SjCcQP-aD5Ca4I07Xa6mU0kUr4hn-uegcG1GOKPieRTnRqwfbu_3qHYyf2PV2cJJkvLWCubkCpEttBFgsCXb0XTwOEHXJUA06sTe4mxIzCkwNpjxN_EebOJhzxA-5tIfMdwjKksL4SR8SysUcd2bqIGrRC7IsKnDA9YsycdR2Up9AbGgVbJDqZoH8FHy0tdprYhHy1g"}'
+		headers="[[_getHeaders()]]"
```
Comment (jbydeley-d2l) [inline activity-exemptions.html]: Why are we adding Bearer tokens? We're in the LMS. We can just call the API.

File: activity-exemptions.html [hunk 4]
```diff
@@ -187,34 +193,18 @@
...
-			mapUserData: function() {
-				const userData = this.rawClasslist.map( (user) => {
-					const isExempt = this.rawExemptions.some( (exemptions) => {
-						return exemptions['UserId'] === Number(user.Identifier);
-					});
-					user.IsExempt = isExempt;
-					return user;
-				});
-				this.toMap(userData);
+			_getHeaders: function() {
```
Comment (jbydeley-d2l) [inline activity-exemptions.html]: Not needed

File: activity-exemptions.html (hunk 3 above)
Comment (jwong-d2l) [reply activity-exemptions.html]: Sorry, I had branched off @lawong-d2l 's branch, so I just moved what was already in there.

File: activity-exemptions.html [hunk 5]
```diff
@@ -72,7 +83,7 @@
 					</tr>
 				</thead>
 				<tbody>
-					<template is="dom-repeat" items="[[userData]]">
+					<template is="dom-repeat" items="[[userData]]" observe="IsExempt">
```
Comment (lawong-d2l) [inline activity-exemptions.html]: Cool stuff. This is new to me.

File: activity-exemptions.html [hunk 6]
```diff
@@ -44,12 +45,22 @@
...
 		on-response="handleExemptionsResponse"
 		method="GET"
 		></iron-ajax>
 
 		<iron-ajax
+		id="createExemption"
+		url="[[createExemptionUrl]]"
+		headers="[[_getHeaders()]]"
+		handle-as="json"
+		on-response="handleIt"
```
Comment (lawong-d2l) [inline activity-exemptions.html]: We can remove this right? Same with the method definition

File: activity-exemptions.html [hunk 7]
```diff
@@ -227,23 +239,21 @@
...
 					.filter( (element) => element.querySelector('.checkbox-user[checked]')
 						&& element.data['IsExempt'] !== isExempt );
-				filteredSelection.forEach( (element) => {
-					const user = this.mappedUserData.get(element.data.Identifier);
-					user.IsExempt = isExempt;
+
+				var allPromises = filteredSelection.map( (element) => {
+					this.createExemptionUrl = `${this.hostUrl}/d2l/api/le/1.26/${this.orgUnitId}/activities/exemptions?userId=${element.data.Identifier}&activityId=${this.activityId}`;
+					return this.$.createExemption.generateRequest().completes.then( ()=> {
+						const rowIndex = element.rowIndex - 1; //offset for header
```
Comment (lawong-d2l) [inline activity-exemptions.html]: rowIndex is neat.

File: activity-exemptions.html [hunk 8]
```diff
@@ -227,23 +239,21 @@
...
 				});
-				// Deep copy of the mappedUserData to notify polymer that value has been changed
-				const mapCopy = JSON.stringify([...this.mappedUserData.entries()]);
-				this.mappedUserData = new Map(JSON.parse(mapCopy));
 
-				this.showSaveToast(isExempt, filteredSelection.length);
-			},
+				Promise.all(allPromises).then(
+					() => {
+						this.showSaveToast(isExempt, allPromises.length);
```
Comment (lawong-d2l) [inline activity-exemptions.html]: Is this still the all or nothing approach for success vs failures for the toast count?

File: activity-exemptions.html [hunk 9]
```diff
@@ -172,33 +196,21 @@
...
 			},
 
 			_getClasslistUrl: function(hostUrl, orgUnitId) {
-				return `https://${hostUrl}/d2l/api/le/1.12/${orgUnitId}/classlist/`;
+				return `${hostUrl}/d2l/api/le/1.12/${orgUnitId}/classlist/`;
 			},
 
 			_getExemptionsUrl: function(hostUrl, orgUnitId, activityId) {
-				return `https://${hostUrl}/d2l/api/le/1.26/${orgUnitId}/activities/exemptions/?userId=&activityId=${activityId}`;
+				return `${hostUrl}/d2l/api/le/1.26/${orgUnitId}/activities/exemptions/?userId=&activityId=${activityId}&showHistory=false`;
```
Comment (lawong-d2l) [inline activity-exemptions.html]: We're adding this default value back in?

Comment (jwong-d2l) [reply activity-exemptions.html]: nope

File: activity-exemptions.html (hunk 8 above)
Comment (jwong-d2l) [reply activity-exemptions.html]: we don't know what to do on error case yet.  Ed will make a story.

**Pull request merged**